# Project 3, Part 1

Kris Barbier 

## Imports and Read in Data

In [14]:
#Imports
import pandas as pd
import numpy as np

In [10]:
#Read in title.basics data set
basics = pd.read_csv('Data/title.basics.tsv', sep='\t', low_memory= False)
basics.head()

tconst titleType            primaryTitle           originalTitle  \
0  tt0000001     short              Carmencita              Carmencita   
1  tt0000002     short  Le clown et ses chiens  Le clown et ses chiens   
2  tt0000003     short          Pauvre Pierrot          Pauvre Pierrot   
3  tt0000004     short             Un bon bock             Un bon bock   
4  tt0000005     short        Blacksmith Scene        Blacksmith Scene   

  isAdult startYear endYear runtimeMinutes                    genres  
0       0      1894      \N              1         Documentary,Short  
1       0      1892      \N              5           Animation,Short  
2       0      1892      \N              4  Animation,Comedy,Romance  
3       0      1892      \N             12           Animation,Short  
4       0      1893      \N              1              Comedy,Short

In [7]:
#Read in title.ratings data set
ratings = pd.read_csv('Data/title.ratings.tsv', sep='\t', low_memory= False)
ratings.head()

tconst  averageRating  numVotes
0  tt0000001            5.7      1990
1  tt0000002            5.8       264
2  tt0000003            6.5      1863
3  tt0000004            5.5       177
4  tt0000005            6.2      2647

In [8]:
#Read in title.akas data set
akas = pd.read_csv('Data/title-akas-us-only.csv', low_memory= False)
akas.head()

titleId  ordering                   title region language        types  \
0  tt0000001         6              Carmencita     US       \N  imdbDisplay   
1  tt0000002         7  The Clown and His Dogs     US       \N           \N   
2  tt0000005        10        Blacksmith Scene     US       \N  imdbDisplay   
3  tt0000005         1     Blacksmithing Scene     US       \N  alternative   
4  tt0000005         6     Blacksmith Scene #1     US       \N  alternative   

              attributes isOriginalTitle  
0                     \N               0  
1  literal English title               0  
2                     \N               0  
3                     \N               0  
4                     \N               0

## Filtering Datasets

### Akas

- Keep only US movies
- Replace "\N" with np.nan

In [13]:
#Check .info() for data types and null values
akas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1452564 entries, 0 to 1452563
Data columns (total 8 columns):
 #   Column           Non-Null Count    Dtype 
---  ------           --------------    ----- 
 0   titleId          1452564 non-null  object
 1   ordering         1452564 non-null  int64 
 2   title            1452564 non-null  object
 3   region           1452564 non-null  object
 4   language         1452564 non-null  object
 5   types            1452564 non-null  object
 6   attributes       1452564 non-null  object
 7   isOriginalTitle  1452564 non-null  object
dtypes: int64(1), object(7)
memory usage: 88.7+ MB


- The "\N" values are not coming up as null values.
- There are only integer and object type columns.

In [12]:
#Check to make sure the US is the only value in region
akas['region'].value_counts()

US    1452564
Name: region, dtype: int64

- US is the only region, so nothing more needs to be done for step 1.

In [15]:
#Replace "\N" with np.nan
akas = akas.replace({'\\N':np.nan})
#Verify the change
akas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1452564 entries, 0 to 1452563
Data columns (total 8 columns):
 #   Column           Non-Null Count    Dtype 
---  ------           --------------    ----- 
 0   titleId          1452564 non-null  object
 1   ordering         1452564 non-null  int64 
 2   title            1452564 non-null  object
 3   region           1452564 non-null  object
 4   language         4018 non-null     object
 5   types            981678 non-null   object
 6   attributes       47016 non-null    object
 7   isOriginalTitle  1451222 non-null  object
dtypes: int64(1), object(7)
memory usage: 88.7+ MB


- All filtering steps are complete for akas data set.

### Basics

- Keep only US movies 
- Replace "\N" with np.nan
- Eliminate movies that are null for runtimeMinutes
- Eliminate movies that are null for genre
- Keep only titleType==Movie
- Convert the startYear column to float data type.
- Filter the dataframe using startYear. Keep years between 2000-2021 (Including 2000 and 2021)
- Eliminate movies that include "Documentary" in the genre

In [16]:
#Check .info() for null values and data types
basics.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10100795 entries, 0 to 10100794
Data columns (total 9 columns):
 #   Column          Dtype 
---  ------          ----- 
 0   tconst          object
 1   titleType       object
 2   primaryTitle    object
 3   originalTitle   object
 4   isAdult         object
 5   startYear       object
 6   endYear         object
 7   runtimeMinutes  object
 8   genres          object
dtypes: object(9)
memory usage: 693.6+ MB


- There are only object type columns, and startYear will need to be changed to a float.
- Null values are not calculated here because there are so many values in the data set.

In [17]:
#Filter the basics table down to only include the US by using the filter akas dataframe
keepers = basics['tconst'].isin(akas['titleId'])
keepers


0            True
1            True
2           False
3           False
4            True
            ...  
10100790    False
10100791    False
10100792    False
10100793    False
10100794    False
Name: tconst, Length: 10100795, dtype: bool

In [21]:
#Filter basics
basics = basics[keepers]
#Verify the change
basics.info()

/var/folders/5v/9gj8mfq115g1d799797d3z5m0000gp/T/ipykernel_83611/1017257114.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  basics = basics[keepers]


<class 'pandas.core.frame.DataFrame'>
Int64Index: 1365339 entries, 0 to 10100750
Data columns (total 9 columns):
 #   Column          Non-Null Count    Dtype 
---  ------          --------------    ----- 
 0   tconst          1365339 non-null  object
 1   titleType       1365339 non-null  object
 2   primaryTitle    1365339 non-null  object
 3   originalTitle   1365339 non-null  object
 4   isAdult         1365339 non-null  object
 5   startYear       1267412 non-null  object
 6   endYear         37378 non-null    object
 7   runtimeMinutes  863322 non-null   object
 8   genres          1336918 non-null  object
dtypes: object(9)
memory usage: 104.2+ MB


In [20]:
#Replace "\N" with np.nan
basics = basics.replace({'\\N':np.nan})
#Verify the change
basics.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1365339 entries, 0 to 10100750
Data columns (total 9 columns):
 #   Column          Non-Null Count    Dtype 
---  ------          --------------    ----- 
 0   tconst          1365339 non-null  object
 1   titleType       1365339 non-null  object
 2   primaryTitle    1365339 non-null  object
 3   originalTitle   1365339 non-null  object
 4   isAdult         1365339 non-null  object
 5   startYear       1267412 non-null  object
 6   endYear         37378 non-null    object
 7   runtimeMinutes  863322 non-null   object
 8   genres          1336918 non-null  object
dtypes: object(9)
memory usage: 104.2+ MB


In [22]:
#Eliminate movies that are null for runtimeMinutes
basics = basics.dropna(subset= ['runtimeMinutes'])
#Verify the change
basics.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 863322 entries, 0 to 10100656
Data columns (total 9 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   tconst          863322 non-null  object
 1   titleType       863322 non-null  object
 2   primaryTitle    863322 non-null  object
 3   originalTitle   863322 non-null  object
 4   isAdult         863322 non-null  object
 5   startYear       854521 non-null  object
 6   endYear         20773 non-null   object
 7   runtimeMinutes  863322 non-null  object
 8   genres          854274 non-null  object
dtypes: object(9)
memory usage: 65.9+ MB


In [24]:
#Eliminate movies that are null for genre
basics = basics.dropna(subset= ['genres'])
#Verify the change
basics.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 854274 entries, 0 to 10100656
Data columns (total 9 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   tconst          854274 non-null  object
 1   titleType       854274 non-null  object
 2   primaryTitle    854274 non-null  object
 3   originalTitle   854274 non-null  object
 4   isAdult         854274 non-null  object
 5   startYear       845546 non-null  object
 6   endYear         20394 non-null   object
 7   runtimeMinutes  854274 non-null  object
 8   genres          854274 non-null  object
dtypes: object(9)
memory usage: 65.2+ MB


In [27]:
#Keep only titleType==Movie
basics = basics[basics['titleType'] == 'movie']
#Verify the change
basics['titleType'].value_counts()

movie    203767
Name: titleType, dtype: int64

In [28]:
#Convert the startYear column to float data type
basics['startYear'] = basics['startYear'].astype(float)
#Verify the change with .info
basics.info()

/var/folders/5v/9gj8mfq115g1d799797d3z5m0000gp/T/ipykernel_83611/3040624165.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  basics['startYear'] = basics['startYear'].astype(float)


<class 'pandas.core.frame.DataFrame'>
Int64Index: 203767 entries, 8 to 10100561
Data columns (total 9 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   tconst          203767 non-null  object 
 1   titleType       203767 non-null  object 
 2   primaryTitle    203767 non-null  object 
 3   originalTitle   203767 non-null  object 
 4   isAdult         203767 non-null  object 
 5   startYear       200238 non-null  float64
 6   endYear         0 non-null       object 
 7   runtimeMinutes  203767 non-null  object 
 8   genres          203767 non-null  object 
dtypes: float64(1), object(8)
memory usage: 15.5+ MB


In [33]:
#Filter the dataframe using startYear. 
#Keep years between 2000-2021 (Including 2000 and 2021)
basics = basics[(basics['startYear'] >= 2000) & (basics['startYear'] <= 2021)]
#Verify the change with .info
basics.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 114436 entries, 34800 to 10100561
Data columns (total 9 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   tconst          114436 non-null  object 
 1   titleType       114436 non-null  object 
 2   primaryTitle    114436 non-null  object 
 3   originalTitle   114436 non-null  object 
 4   isAdult         114436 non-null  object 
 5   startYear       114436 non-null  float64
 6   endYear         0 non-null       object 
 7   runtimeMinutes  114436 non-null  object 
 8   genres          114436 non-null  object 
dtypes: float64(1), object(8)
memory usage: 8.7+ MB


In [34]:
#Eliminate movies that include "Documentary" in the genre
is_documentary = basics['genres'].str.contains('documentary',case=False)
basics = basics[~is_documentary]
#Verify the changes with .info
basics.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 81890 entries, 34800 to 10100561
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   tconst          81890 non-null  object 
 1   titleType       81890 non-null  object 
 2   primaryTitle    81890 non-null  object 
 3   originalTitle   81890 non-null  object 
 4   isAdult         81890 non-null  object 
 5   startYear       81890 non-null  float64
 6   endYear         0 non-null      object 
 7   runtimeMinutes  81890 non-null  object 
 8   genres          81890 non-null  object 
dtypes: float64(1), object(8)
memory usage: 6.2+ MB


### Ratings

- Keep only movies that were included in your final title basics dataframe above.
- Replace "\N" with np.nan (if any)

In [35]:
#Check .info for column information
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1340671 entries, 0 to 1340670
Data columns (total 3 columns):
 #   Column         Non-Null Count    Dtype  
---  ------         --------------    -----  
 0   tconst         1340671 non-null  object 
 1   averageRating  1340671 non-null  float64
 2   numVotes       1340671 non-null  int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 30.7+ MB


In [36]:
#Keep only movies that were included in your final title basics dataframe
keepers = ratings['tconst'].isin(basics['tconst'])
keepers

0          False
1          False
2          False
3          False
4          False
           ...  
1340666    False
1340667    False
1340668    False
1340669    False
1340670    False
Name: tconst, Length: 1340671, dtype: bool

In [37]:
#Filter basics
ratings = ratings[keepers]
#Verify the change
ratings.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 68042 entries, 17882 to 1340641
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   tconst         68042 non-null  object 
 1   averageRating  68042 non-null  float64
 2   numVotes       68042 non-null  int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 2.1+ MB


In [39]:
#Replace "\N" with np.nan (if any)
ratings = ratings.replace({'\\N':np.nan})
#Verify the change
ratings.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 68042 entries, 17882 to 1340641
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   tconst         68042 non-null  object 
 1   averageRating  68042 non-null  float64
 2   numVotes       68042 non-null  int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 2.1+ MB


## Final Steps

In [40]:
#Save final .info for basics
basics.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 81890 entries, 34800 to 10100561
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   tconst          81890 non-null  object 
 1   titleType       81890 non-null  object 
 2   primaryTitle    81890 non-null  object 
 3   originalTitle   81890 non-null  object 
 4   isAdult         81890 non-null  object 
 5   startYear       81890 non-null  float64
 6   endYear         0 non-null      object 
 7   runtimeMinutes  81890 non-null  object 
 8   genres          81890 non-null  object 
dtypes: float64(1), object(8)
memory usage: 6.2+ MB


- There are 81,890 movies remaining in this dataframe. There are no null values in any column, except endYear, which is entirely null, and can likely be dropped in later steps.

In [41]:
#Save final .info for ratings
ratings.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 68042 entries, 17882 to 1340641
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   tconst         68042 non-null  object 
 1   averageRating  68042 non-null  float64
 2   numVotes       68042 non-null  int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 2.1+ MB


- There are 68,042 movies remaining in this dataframe, with no null values.

In [42]:
#Save final .info for akas
akas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1452564 entries, 0 to 1452563
Data columns (total 8 columns):
 #   Column           Non-Null Count    Dtype 
---  ------           --------------    ----- 
 0   titleId          1452564 non-null  object
 1   ordering         1452564 non-null  int64 
 2   title            1452564 non-null  object
 3   region           1452564 non-null  object
 4   language         4018 non-null     object
 5   types            981678 non-null   object
 6   attributes       47016 non-null    object
 7   isOriginalTitle  1451222 non-null  object
dtypes: int64(1), object(7)
memory usage: 88.7+ MB


- There are 1,452,564 movies remaining in this dataframe. Four columns have missing values.

In [43]:
#Save basics as csv.gz file
basics.to_csv("Data/title_basics.csv.gz",compression='gzip',index=False)

In [44]:
#Open saved file and preview again
basics = pd.read_csv("Data/title_basics.csv.gz", low_memory = False)
basics.head()

tconst titleType                                       primaryTitle  \
0  tt0035423     movie                                     Kate & Leopold   
1  tt0062336     movie  The Tango of the Widower and Its Distorting Mi...   
2  tt0068865     movie                                Lives of Performers   
3  tt0069049     movie                         The Other Side of the Wind   
4  tt0088751     movie                                  The Naked Monster   

                               originalTitle  isAdult  startYear  endYear  \
0                             Kate & Leopold        0     2001.0      NaN   
1  El tango del viudo y su espejo deformante        0     2020.0      NaN   
2                        Lives of Performers        0     2016.0      NaN   
3                 The Other Side of the Wind        0     2018.0      NaN   
4                          The Naked Monster        0     2005.0      NaN   

   runtimeMinutes                  genres  
0             118  Comedy,Fantasy,Romance  
1              70                   Drama  
2              90                   Drama  
3             122                   Drama  
4             100    Comedy,Horror,Sci-Fi

In [45]:
#Save ratings as csv.gz file
ratings.to_csv("Data/title_ratings.csv.gz",compression='gzip',index=False)

In [46]:
#Open saved file and preview again
ratings = pd.read_csv("Data/title_ratings.csv.gz", low_memory = False)
ratings.head()

tconst  averageRating  numVotes
0  tt0035423            6.4     87357
1  tt0062336            6.4       179
2  tt0068865            5.4        75
3  tt0069049            6.7      7799
4  tt0088751            5.2       339

In [47]:
#Save akas as csv.gz file
akas.to_csv("Data/akas.csv.gz",compression='gzip',index=False)

In [48]:
#Open saved file and preview again
akas = pd.read_csv("Data/akas.csv.gz", low_memory = False)
akas.head()

titleId  ordering                   title region language        types  \
0  tt0000001         6              Carmencita     US      NaN  imdbDisplay   
1  tt0000002         7  The Clown and His Dogs     US      NaN          NaN   
2  tt0000005        10        Blacksmith Scene     US      NaN  imdbDisplay   
3  tt0000005         1     Blacksmithing Scene     US      NaN  alternative   
4  tt0000005         6     Blacksmith Scene #1     US      NaN  alternative   

              attributes  isOriginalTitle  
0                    NaN              0.0  
1  literal English title              0.0  
2                    NaN              0.0  
3                    NaN              0.0  
4                    NaN              0.0